### PR TITLE
issue_44 handling Canvas paging and API threshold

### PR DESCRIPTION
**Pull request by zqian**
major changes include:

1. handling of Canvas and ESB API call limit
2. handling of Canvas API call paging
3. removed the Canvas API call to "#{$canvasUrl}/api/v1/sections/#{sectionValue}", getting section information from the "#{$canvasUrl}/api/v1/courses/#{courseId}/sections" call instead.

### CODE REVIEW

File: SIS_import/sis_set_url.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@
 require "json"
 require "fileutils"
 require "nokogiri"
+require "nokogiri"
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: Once is probably enough

File: SIS_import/sis_set_url.rb [hunk 2]
```diff
@@ -11,6 +12,54 @@
 require "uri"
 require "time"
 
+# variables for ESB API call throttling
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: This and canvas throttle are nearly identical.  Make a separate class for clarity and to avoid typos and extra maintenance..

File: SIS_import/sis_set_url.rb [hunk 3]
```diff
@@ -11,6 +12,54 @@
...
+		#rest the esb call count
+		@esb_call_count = 0
+	end
+end
+
+# variables for Canva API call throttling
+@Canvas_call_count = 0
+@Canvas_start_time = Time.now
+@Canvas_time_interval_in_seconds = 3600
+@Canvas_end_time = @esb_start_time + @esb_time_interval_in_seconds # one minute apart
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: Canvas or ESB?

File: SIS_import/sis_set_url.rb [hunk 4]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+	json = parse_canvas_API_response_json(url, response, json_attribute, outputFile)
+
+	# array of page urls if any
+	page_urls = get_all_canvas_page_urls(response.headers)
+	if (!page_urls.nil?)
+		# there is paging involved
+		# need to make further API calls
+		# will concat the result json arrays
+		page_urls.each do |url|
+			response = actual_Canvas_API_call(url, authorization_string, outputFile)
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: what happens if one call fails?

File: SIS_import/sis_set_url.rb [hunk 5]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+	page_url_part_one = nil
+	# page url part two
+	page_url_part_two = nil
+
+	link_header = response_headers[:link]
+	link_page_urls = link_header.split(',')
+	link_page_urls.each do |page_link|
+		if page_link.include? "rel=\"last\""
+			# this is the last link, get the page id]
+			# get the param from url string
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: make this clause into a method

File: SIS_import/sis_set_url.rb [hunk 6]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+		end
+	end
+	outputFile.write("Canvas API call JSON \n #{json}\n")
+
+	return json
+end
+
+## the get call to Canvas
+def actual_Canvas_API_call(url, authorization_string, outputFile)
 	response = RestClient.get url, {:Authorization => "Bearer #{authorization_string}",
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: What happens if an error is found?

File: SIS_import/sis_set_url.rb [hunk 7]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+	return response
+end
+
+## parse the response object into json object
+def parse_canvas_API_response_json(url, response, json_attribute, outputFile)
+
+	p "Canvas API call " << url
+	outputFile.write("Canvas API call " << url << "\n")
+
+	json = JSON.parse(response)
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: What happens if parse fails?

File: SIS_import/sis_set_url.rb [hunk 8]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+	# default last page number to be 1
+	last_page_number = 1
+	# page url part one
+	page_url_part_one = nil
+	# page url part two
+	page_url_part_two = nil
+
+	link_header = response_headers[:link]
+	link_page_urls = link_header.split(',')
+	link_page_urls.each do |page_link|
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: May be clearer to use detect rather than each.

File: SIS_import/sis_set_url.rb [hunk 9]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+		if page_link.include? "rel=\"last\""
+			# this is the last link, get the page id]
+			# get the param from url string
+			last_page_url_array= page_link.split(';')
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: make this clause a method.

File: SIS_import/sis_set_url.rb [hunk 10]
```diff
@@ -74,11 +242,14 @@ def ESB_APICall(url, authorization_string, content_type, request_type, param_has
 	sock.cert_store = store
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: why use explicit cert store?

File: SIS_import/sis_set_url.rb [hunk 11]
```diff
@@ -74,11 +242,14 @@ def ESB_APICall(url, authorization_string, content_type, request_type, param_has
...
-	sock.open_timeout = 30
+	sock.read_timeout = 60
+	sock.open_timeout = 60
 	sock.start do |http|
 		response = http.request(request)
 	end
+
+	p response
+
 	return JSON.parse(response.body)
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: What if there is a parse error?

File: SIS_import/sis_set_url.rb [hunk 12]
```diff
@@ -117,8 +290,11 @@ def getMPathwayTerms(esbToken)
 def processTermCourses(mPathwayTermSet,esbToken, outputFile)
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: Avoid most nested each by using methods.  Avoid nesting by using quick return/next if possible.

File: SIS_import/sis_set_url.rb [hunk 13]
```diff
@@ -140,28 +329,33 @@ def processTermCourses(mPathwayTermSet,esbToken, outputFile)
...
+									if (result_json.has_key? "setLMSURLResponse")
+										message = "set url result for section id=#{sectionParsedSISID} with Canvas courseId=#{courseId}: result status=#{result_json["setLMSURLResponse"]["Resultcode"]} and result message=#{result_json["setLMSURLResponse"]["ResultMessage"]}"
+									else
+										message = "set url result for section id=#{sectionParsedSISID} with Canvas courseId=#{courseId}: result #{result_json.to_s}"
 									end
+									# write into output file
+									outputFile.write(message + "\n")
 								end
 							}
 						end
```
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: Why does processTermCourses say canvas API but require esbToken?

Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: Put ARGV processing in methods.

Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: 496, 499: Most likely don't need to put code after an abort or exit in an else clause.

File: SIS_import/sis_set_url.rb [hunk 14]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+			response = actual_Canvas_API_call(url, authorization_string, outputFile)
+
+			json_paging_data = parse_canvas_API_response_json(url, response, json_attribute, outputFile)
+			# merge in all element in the paging data array
+			json_paging_data.each do |json_paging_data_element|
+				json.push json_paging_data_element
+			end
+		end
+	end
+	outputFile.write("Canvas API call JSON \n #{json}\n")
```
Comment (bkirschn) [inline SIS_import/sis_set_url.rb]: Should this be for debugging only? If always written, is the output file periodically purged?

File: SIS_import/sis_set_url.rb [hunk 15]
```diff
@@ -24,15 +73,134 @@ def refreshESBToken()
...
+## the get call to Canvas
+def actual_Canvas_API_call(url, authorization_string, outputFile)
 	response = RestClient.get url, {:Authorization => "Bearer #{authorization_string}",
 	                                :accept => "application/json",
 	                                :verify_ssl => true}
-return JSON.parse(response)
+
+	@Canvas_call_count = @Canvas_call_count + 1
+	p "Canvas call #{@Canvas_call_count}"
+	outputFile.write "Canvas call #{@Canvas_call_count}"
```
Comment (bkirschn) [inline SIS_import/sis_set_url.rb]: Given throttling issues, I wonder if a timestamp would be useful?

File: SIS_import/sis_set_url.rb (hunk 13 above)
Comment (dlhaines) [inline SIS_import/sis_set_url.rb]: My main concerns are:
- Hard to follow because of nested if clauses and loops.
- Error handling is unclear because of being hard to follow, and because of a lack of rescue clauses.  E.g. what happens if there is a json parse error?  Are all updates lost? Just some?  Just one?  How can it be restarted to pick up after an error?

File: SIS_import/sis_set_url.rb [hunk 16]
```diff
@@ -117,8 +290,11 @@ def getMPathwayTerms(esbToken)
 def processTermCourses(mPathwayTermSet,esbToken, outputFile)
 
 	term_data = Canvas_API_call("#{$canvasUrl}/api/v1/accounts/1/terms",
-	                           $canvasToken)
-	term_data["enrollment_terms"].each {|term|
+	                            {:per_page => $page_size},
+	                            $canvasToken,
+															"enrollment_terms",
+															outputFile)
```
Comment (bkirschn) [inline SIS_import/sis_set_url.rb]: tab vs space?

File: SIS_import/sis_set_url.rb [hunk 17]
```diff
@@ -126,13 +302,26 @@ def processTermCourses(mPathwayTermSet,esbToken, outputFile)
...
+
 			outputFile.write("for term SIS_ID=#{term["sis_term_id"]} and Canvas term id=#{termId}\n")
 
 			# Web Service call
-			json_data = Canvas_API_call("#{$canvasUrl}/api/v1/accounts/1/courses?per_page=#{$page_size}&enrollment_term_id=#{termId}&published=true&with_enrollments=true",
-			                  $canvasToken)
-			#outputFile.write("#{json_data}\n")
+			json_data = Canvas_API_call("#{$canvasUrl}/api/v1/accounts/1/courses",
+			                            { :enrollment_term_id=>termId,
+																		:published =>true,
```
Comment (bkirschn) [inline SIS_import/sis_set_url.rb]: spacing?

File: SIS_import/sis_set_url.rb [hunk 18]
```diff
@@ -140,28 +329,33 @@ def processTermCourses(mPathwayTermSet,esbToken, outputFile)
...
+									# we will use just the last 5-digit of the section id
+									sectionParsedSISID = sectionParsedSISID[4,8]
+									result_json = setMPathwayUrl($canvasUrl, $esbUrl, esbToken, sisTermId, sectionParsedSISID, courseId)
+									if (result_json.has_key? "setLMSURLResponse")
+										message = "set url result for section id=#{sectionParsedSISID} with Canvas courseId=#{courseId}: result status=#{result_json["setLMSURLResponse"]["Resultcode"]} and result message=#{result_json["setLMSURLResponse"]["ResultMessage"]}"
+									else
+										message = "set url result for section id=#{sectionParsedSISID} with Canvas courseId=#{courseId}: result #{result_json.to_s}"
 									end
+									# write into output file
+									outputFile.write(message + "\n")
```
Comment (bkirschn) [inline SIS_import/sis_set_url.rb]: Do we need some automated way to monitor this outputFile for errors?

File: SIS_import/sis_set_url.rb (hunk 1 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: removed the redundant import.

File: SIS_import/sis_set_url.rb (hunk 3 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: Canvas :(

File: SIS_import/sis_set_url.rb (hunk 14 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: commented out. Should be debug only.
As for the output file, there is one file generated for each script run, named after "Canvas_set_url_<timestamp>.txt"

File: SIS_import/sis_set_url.rb (hunk 17 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: spacing fixed

File: SIS_import/sis_set_url.rb (hunk 8 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: used detect instead.

File: SIS_import/sis_set_url.rb (hunk 2 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: Done. Use the same function for both Canvas and ESB call throttles.

File: SIS_import/sis_set_url.rb (hunk 6 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: added rescue statement to log error into log file

File: SIS_import/sis_set_url.rb (hunk 15 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: Done. Timestamp added.

File: SIS_import/sis_set_url.rb (hunk 7 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: rescue statement in place, log error into log file.

File: SIS_import/sis_set_url.rb (hunk 11 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: added rescue statement

File: SIS_import/sis_set_url.rb (hunk 16 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: fixed.

File: SIS_import/sis_set_url.rb (hunk 13 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: removed esbToken from processTermCourses param list.

File: SIS_import/sis_set_url.rb (hunk 18 above)
Comment (zqian) [inline SIS_import/sis_set_url.rb]: errors can now be sent via email to ENV['MAILTO'] email settings.
